cmd/snap-confine/seccomp-support.c

```diff
@@ -109,12 +135,94 @@ static void validate_bpfpath_is_safe(const char *path)
 	}
 }
 
+static void sc_cleanup_sock_fprog(struct sock_fprog *prog)
+{
+	free(prog->filter);
+	prog->filter = NULL;
+}
+
+static void sc_must_read_filter_from_file(FILE *file, uint32_t len_bytes,
+					  char *what, struct sock_fprog *prog)
+{
+	if (len_bytes == 0) {
+		die("%s filter may only be empty in unrestricted profiles",
+		    what);
+	}
+	prog->len = len_bytes / sizeof(struct sock_filter);
+	prog->filter = (struct sock_filter *)malloc(len_bytes);
+	if (prog->filter == NULL) {
+		die("cannot allocate %u bytes of memory for %s seccomp filter ",
+		    len_bytes, what);
+	}
+	size_t num_read = fread(prog->filter, 1, len_bytes, file);
+	if (ferror(file)) {
+		die("cannot read %s filter", what);
+	}
+	if (num_read != len_bytes) {
+		die("short read for filter %s %zu != %i", what, num_read,
+		    len_bytes);
+	}
+}
+
+static FILE *sc_must_read_and_validate_header_from_file(const char
+							*profile_path, struct
+							sc_seccomp_file_header
+							*hdr)
+{
+	FILE *file = fopen(profile_path, "rb");
```

Comment:
Nitpick, if we move the fopen outside the code becomes easier to reason about IMO.

Reply:
Umm, I've mixed feelings with this one. I would also like to move the fopen outside, however, that will force us to either move the complete following code block outside, pass profile_path argument in addition to the new FILE argument or provide less verbose error messages. I don't really know which option is best here tbh

```
	FILE *file = fopen(profile_path, "rb");
	if (file == NULL) {
		die("cannot open seccomp filter %s", profile_path);
	}
	size_t num_read =
	    fread(hdr, 1, sizeof(struct sc_seccomp_file_header), file);
	if (ferror(file) != 0) {
		die("cannot read seccomp profile %s", profile_path);
	}
	if (num_read < sizeof(struct sc_seccomp_file_header)) {
		die("short read on seccomp header: %zu", num_read);
	}
```

Reply:
I just want to split the responsibility: handling the file is outside is easy, no complex life-cycle to worry about. Once you have the file you read the header from a FILE that's not yours and then you read the body. In all the cases lifecycle of the file is trivial. The reading functions should be exactly what you have now but FILE should not be returned.

Reply:
Done in 2ae52cbfa2d8635f7365cbe8e21ad31dd4a02f57